Close #600. Allow devs to test list of interviews on their server.

plocket opened the pull request:
Closes #600.

plocket committed:
Close #600. Allow devs to test list of interviews on their server.

BryceStevenWilley commented on CHANGELOG.md:
```diff
@@ -38,7 +38,9 @@ Format:
-<!-- ## [Unreleased] -->
+## [Unreleased]
+### Changed
+- Allow devs to test list of interviews on their server. See https://github.com/SuffolkLITLab/ALKiln/issues/600.
```
I would say that this lets you test installed interviews, right? Nothing in this PR is specific to getting the full list and testing it.

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -752,10 +752,27 @@ module.exports = {
-    if ( session_vars.get_debug() ) { console.log( interview_url ); }
+    let interview_url = null;
+    // If the file is already a full url, just use that arbitrary url
+    if ( file_name.includes( `http` ) ) {
```
I would make this more specific. Not sure if JS has a startswith, but something like `file_name.startswith(`http://`) || file_name.startswith(`https://`). Just so we don't have any false positives.

BryceStevenWilley commented on lib/scope.js:
```diff
@@ -794,35 +813,55 @@ module.exports = {
       scope.page.waitForSelector( `#daMainQuestion`, { timeout: scope.timeout }),
       scope.page.waitForSelector( `#da-retry`, { timeout: scope.timeout }),
-    ]);
+    ]).catch( function ( error ) {
```
probably out of scope for this PR, but for my own curiosity:

Can we tell when the page is completely finished loading? Would mean that we wouldn't have to wait for the timeout when everything is done and sitting there.

plocket replied:
We already do that by checking for any of those elements to be present. As soon as one is there, the race stops and the code moves on. The timeout is just in case none of them are found. We don't want to wait for them for forever. Of course, we may be missing some kind of indicator of a page having loaded, but I'm not sure how to track that down.

BryceStevenWilley replied:
> The timeout is just in case none of them are found. We don't want to wait for them for forever. 

That was my question, if none of them appear, we just wait for the timeout, which is a little silly IMO, especially for tests with longer timeouts.

This stackoverflow has some code that we should consider using: https://stackoverflow.com/questions/52497252/puppeteer-wait-until-page-is-completely-loaded.

plocket replied:
Always good to have a new eye on things! When I looked, I only remember finding stuff about `networidle` being unreliable. This looks like an interesting solution. We should back it up with the timeout as well, though, just in case it keeps changing for some reason. See [the comment below the answer](https://stackoverflow.com/questions/52497252/puppeteer-wait-until-page-is-completely-loaded#comment118988927_61304202) for an example.

[Do we already have an issue for this? ~I think I added one.~ That was a different issue about a different aspect of this.]

BryceStevenWilley reviewed (approved):
LGTM! I think it should be described differently in the CHANGELOG, but the changes themselves look great

plocket committed:
Update test and don't print possible secrets

plocket committed:
Get more specific with url str test and CHANGELOG, but need to

document that arbitrary urls won't be secret in the report -
they'll be printed as is. We can't detect what's secret in there.

Also fixed new tests texts.

plocket commented:
Documentation: arbitrary url won't keep secrets secret

Arbitrary urls won't be secret in the report - they'll be printed as is. We can't detect what's secret in there.

Make this into an issue on the documentation site.

plocket committed:
Update text of arbitrary url tests

plocket committed:
Remove possibly secret server address from report

plocket commented:
I decided to instead take the possibly secret url out of the report. The arbitrary url may have the server address, which is possibly given as a secret and thus would be expected to remain secret. The only reason it's a secret is that there's no other way to pass in an environment variable on GitHub. That said, there'd be no way to differentiate between a secret env var and a non-secret env var anyway.
